sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java

```diff
@@ -531,33 +532,61 @@ public static SharedKeyCredential getSharedKeyCredential(HttpPipeline httpPipeli
      * @throws RuntimeException When I/O error occurs.
      */
     public static Flux<ByteBuffer> convertStreamToByteBuffer(InputStream data, long length, int blockSize) {
+        Pair pair = new Pair();
         final long[] currentTotalLength = new long[1];
-        return Flux.range(0, (int) Math.ceil((double) length / (double) blockSize))
-            .map(i -> i * blockSize)
-            .concatMap(pos -> Mono.fromCallable(() -> {
-                long count = pos + blockSize > length ? length - pos : blockSize;
-                byte[] cache = new byte[(int) count];
-                int lastIndex = data.read(cache);
-                currentTotalLength[0] += lastIndex;
-                if (lastIndex < count) {
+        return Flux.just(true)
+            .repeat()
+            .map(ignore -> {
+                byte[] buffer = new byte[blockSize];
+                try {
+                    int numBytes = data.read(buffer);
```

Comment:
Now that we've repeat() and `length` I guess we can simplify this a bit and avoid extra allocation of `Pair` type & avoid extra filter , onComplete and flatmap. We could also move validations to within this map. I didn't test below code but this is something in my mind.

```java
public static Flux<ByteBuffer> convertStreamToByteBuffer(InputStream data, long length, int blockSize) {
    final long[] currentTotalLength = new long[1];
    return Flux.just(true)
        .repeat()
        .map(ignore -> {
             byte[] buffer = new byte[blockSize];
             int totalBytesRead = 0;
             while (totalBytesRead < blockSize) {
                 int numBytesRead = data.read(buffer, totalBytesRead, blockSize - totalBytesRead);
                 if (numBytesRead == -1)  {
                     break;
                 } else {
                    totalBytesRead += numBytesRead; 
                 }
             }
             if (totalBytesRead == 0) {
                 // If totalBytesRead is zero after the above loop then means either we reached EOF
                 // or blockSize is zero
                 if (currentTotalLength[0] < length) {
                     throw "Stream has less than provided length!"
                 } else {
                    // Emit an empty ByteBuffer that get filtered out by below takeUntil.
                    return ByteBuffer.wrap(buffer, 0, 0);
                 }
             } else {
                 currentTotalLength[0] += totalBytesRead;
                 if (currentTotalLength[0] > length) {
                     throw "Stream has more than provided length!"
                 } else {
                    return ByteBuffer.wrap(buffer, 0, totalBytesRead);
                 }
             }
         })
         .takeUntil(bb -> !bb.hasRemaining());
```

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/5754

An issue to track the simplification.